src/HAL/Hal.c

```diff
@@ -0,0 +1,314 @@
+Hal_uart_init_uart0_pio(Pio_Port* const port, Pio_Port_Config* const pioConfigTx, Pio_Port_Config* const pioConfigRx)
+{
+    *port = Pio_Port_A;
+
+    pioConfigRx->pins = PIO_PIN_9;
+    pioConfigRx->pinsConfig.control = Pio_Control_PeripheralA;
+
+    pioConfigTx->pins = PIO_PIN_10;
+    pioConfigTx->pinsConfig.control = Pio_Control_PeripheralA;
+}
+inline static void
+Hal_uart_init_uart1_pio(Pio_Port* const port, Pio_Port_Config* const pioConfigTx, Pio_Port_Config* const pioConfigRx)
+{
+    *port = Pio_Port_A;
+
+    pioConfigRx->pins = PIO_PIN_5;
+    pioConfigRx->pinsConfig.control = Pio_Control_PeripheralC;
+
+    pioConfigTx->pins = PIO_PIN_6;
+    pioConfigTx->pinsConfig.control = Pio_Control_PeripheralC;
+}
+inline static void
+Hal_uart_init_uart2_pio(Pio_Port* const port, Pio_Port_Config* const pioConfigTx, Pio_Port_Config* const pioConfigRx)
+{
+    *port = Pio_Port_D;
+
+    pioConfigRx->pins = PIO_PIN_25;
+    pioConfigRx->pinsConfig.control = Pio_Control_PeripheralC;
+
+    pioConfigTx->pins = PIO_PIN_26;
+    pioConfigTx->pinsConfig.control = Pio_Control_PeripheralC;
+}
+inline static void
+Hal_uart_init_uart3_pio(Pio_Port* const port, Pio_Port_Config* const pioConfigTx, Pio_Port_Config* const pioConfigRx)
+{
+    *port = Pio_Port_D;
+
+    pioConfigRx->pins = PIO_PIN_28;
+    pioConfigRx->pinsConfig.control = Pio_Control_PeripheralA;
+
+    pioConfigTx->pins = PIO_PIN_30;
+    pioConfigTx->pinsConfig.control = Pio_Control_PeripheralA;
+}
+inline static void
+Hal_uart_init_uart4_pio(Pio_Port* const port, Pio_Port_Config* const pioConfigTx, Pio_Port_Config* const pioConfigRx)
+{
+    *port = Pio_Port_D;
+
+    pioConfigRx->pins = PIO_PIN_18;
+    pioConfigRx->pinsConfig.control = Pio_Control_PeripheralC;
+
+    pioConfigTx->pins = PIO_PIN_19;
+    pioConfigTx->pinsConfig.control = Pio_Control_PeripheralC;
+}
+
+inline static void
+Hal_uart_init_pmc(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioA);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart0);
+            break;
+        case Uart_Id_1:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioA);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart1);
+            break;
+        case Uart_Id_2:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart2);
+            break;
+        case Uart_Id_3:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart3);
+            break;
+        case Uart_Id_4:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart4);
+            break;
+    }
+}
+
+inline inline static void
+Hal_uart_init_nvic(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            Nvic_enableInterrupt(Nvic_Irq_Uart0);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart0, 5);
```

Comment:
5 is a magic number. It should be named. Also, why 5?

Reply:
Ok, I will define "5" as a UART priority. 
I have chosen "5" as an interrupt level because all non-kernel interrupts should be smaller (which means a higher number in ARM M7 world) than kernel interrupt priority. The minimal kernel interrupt has a value of 4 next interrupt level is 5 (here I could use configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY + 1 instead of definition). 
There are eight levels of interrupt priorities on ARM M7. I have decided to leave four of them for kernel and four for non-kernel stuff.